client/jetpack-cloud/sections/partner-portal/hooks/index.tsx

```diff
@@ -0,0 +1,342 @@
+	}, [ flag ] );
+
+	return wasTrue;
+}
+
+/**
+ * Handle cursor-based pagination.
+ *
+ * @todo use this in payment method pagination.
+ */
+export function useCursorPagination(
+	enabled: boolean,
+	hasMore: boolean,
+	onNavigateCallback: ( page: number, direction: 'next' | 'prev' ) => void
+): [ number, boolean, ( page: number ) => void ] {
+	const [ page, setPage ] = useState( 1 );
+	const showPagination = usePermanentFlag( hasMore );
+	const onNavigate = useCallback(
+		( newPage: number ) => {
+			const direction = newPage > page ? 'next' : 'prev';
+
+			if ( ! enabled ) {
+				return;
+			}
+
+			if ( newPage < 1 ) {
+				return;
+			}
+
+			if ( direction === 'next' && ! hasMore ) {
+				return;
+			}
+
+			setPage( newPage );
+			onNavigateCallback?.( newPage, direction );
+		},
+		[ enabled, hasMore, setPage, onNavigateCallback ]
+	);
+
+	return [ page, showPagination, onNavigate ];
+}
+
+export function useAssignLicenses(
+	licenseKeys: Array< string >,
+	selectedSite: { ID: number; domain: string } | null
+): [ () => void, boolean ] {
+	const products = useProductsQuery();
+	const dispatch = useDispatch();
+	const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+	const assignLicense = useAssignLicenseMutation( {
+		onError: ( error: Error ) => {
+			dispatch( errorNotice( error.message, { isPersistent: true } ) );
+		},
+	} );
+	const isLoading = assignLicense.isLoading;
+	const selectedSiteId = selectedSite?.ID as number;
+	const assignMultipleLicenses = useCallback( async () => {
+		dispatch(
+			recordTracksEvent( 'calypso_partner_portal_assign_multiple_licenses_submit', {
+				products: licenseKeys.join( ',' ),
+				selected_site: selectedSiteId,
+			} )
+		);
+
+		const keysWithProductNames = licenseKeys
+			.map( ( key ) => {
+				const productSlug = getProductSlugFromLicenseKey( key );
+				const selectedProduct = products?.data?.find?.( ( p ) => p.slug === productSlug );
+
+				return {
+					key,
+					product: selectedProduct,
+				};
+			} )
+			// If we can't determine which product a license is meant for, filter it out
+			.filter( ( { product } ) => Boolean( product ) )
+			// We only need the product's title/display name
+			.map( ( { key, product } ) => ( {
+				key,
+				name: getProductTitle( ( product as APIProductFamilyProduct ).name ),
+			} ) );
+
+		// Purposely catch any error responses and let them through,
+		// so we can pass all their information along as ProductInfo objects
+		// (not currently available via PromiseSettledResult<T>)
+		const apiRequests = keysWithProductNames.map( ( { key, name } ) =>
+			assignLicense
+				.mutateAsync( { licenseKey: key, selectedSite: selectedSiteId } )
+				.then( () => ( { key, name, status: 'fulfilled' } as ProductInfo ) )
+				.catch( () => ( { key, name, status: 'rejected' } as ProductInfo ) )
+		);
+
+		const assignLicenseStatus = {
+			selectedSite: selectedSite?.domain || '',
+			selectedProducts: await Promise.all( apiRequests ),
```

Comment:
I think it is better to use `Promise.allSettled()` here since we want to wait for the requests to be complete.